File: Sorting Plateau/Counting Sort_Merge Sort_Quickselect/Q2. Sort an Array/Q2_SortanArray.java
class Solution {

    // Main function called by LeetCode
    public int[] sortArray(int[] nums) {

        // Call merge sort on the full array
        mergeSort(nums, 0, nums.length - 1);

        // Return the sorted array
        return nums;
    }

    // Function to merge two sorted subarrays
    // arr[l..m] and arr[m+1..r]
    public static void mergeFun(int[] arr, int l, int m, int r) {

        // Number of elements in left subarray
        int n1 = m + 1 - l;

        // Number of elements in right subarray
        int n2 = r - m;

        // Temporary array to store left half
        int[] left = new int[n1];

        // Copy left half elements
        for (int i = 0; i < n1; i++) {
            left[i] = arr[l + i];
        }

        // Temporary array to store right half
        int[] right = new int[n2];

        // Copy right half elements
        for (int i = 0; i < n2; i++) {
            right[i] = arr[m + 1 + i];
        }

        // i -> index for left array
        // j -> index for right array
        // k -> index for original array
        int i = 0, j = 0, k = l;

        // Merge both arrays back into arr
        while (i < n1 || j < n2) {

            // If right array is exhausted
            // OR left element is smaller than right element
            if (j == n2 || (i < n1 && left[i] < right[j])) {

                // Place left element into original array
                arr[k++] = left[i++];

            } else {

                // Place right element into original array
                arr[k++] = right[j++];
            }
        }
    }

    // Recursive merge sort function
    public static void mergeSort(int[] arr, int low, int high) {

        // Continue only if there are at least two elements
        if (low < high) {

            // Find middle index safely
            int middle = (high - low) / 2 + low;

            // Sort left half
            mergeSort(arr, low, middle);

            // Sort right half
            mergeSort(arr, middle + 1, high);

            // Merge the sorted halves
            mergeFun(arr, low, middle, high);
        }
    }
}

File: Sorting Plateau/Counting Sort_Merge Sort_Quickselect/Q2. Sort an Array/Q2_SortanArray.py
class Solution(object):

    # Function to merge two sorted subarrays:
    # A[start .. mid] and A[mid+1 .. end]
    def merge(self, A, start, mid, end, buff):

        left, right = start, mid + 1     # Pointers for left and right subarrays
        s = end - start + 1              # Total number of elements to merge

        # Merge elements into buffer array
        for i in range(s):

            i0 = start + i               # Actual index in buffer / array

            # If left subarray is exhausted
            if left > mid:
                buff[i0] = A[right]      # Take element from right subarray
                right += 1               # Move right pointer

            # If right subarray is exhausted
            elif right > end:
                buff[i0] = A[left]       # Take element from left subarray
                left += 1                # Move left pointer

            # If left element is smaller
            elif A[left] < A[right]:
                buff[i0] = A[left]       # Copy left element
                left += 1                # Move left pointer

            # If right element is smaller or equal
            else:
                buff[i0] = A[right]      # Copy right element
                right += 1               # Move right pointer

        # Copy merged elements back to original array
        for i in range(start, start + s):
            A[i] = buff[i]

    # Recursive merge sort function
    def mergeSort(self, A, start, end, buff):

        # Base case: single element or invalid range
        if end <= start:
            return

        # Find middle index
        mid = start + (end - start) // 2

        # Sort left half
        self.mergeSort(A, start, mid, buff)

        # Sort right half
        self.mergeSort(A, mid + 1, end, buff)

        # Merge both sorted halves
        self.merge(A, start, mid, end, buff)

    # Function called by LeetCode
    def sortArray(self, nums):
        """
        :type nums: List[int]
        :rtype: List[int]
        """

        # Temporary buffer array to avoid repeated allocations
        buff = [0] * len(nums)

        # Apply merge sort on entire array
        self.mergeSort(nums, 0, len(nums) - 1, buff)

        # Return sorted array
        return nums
